azure-kusto-ingest/source/statusQ.js

```diff
@@ -48,117 +46,86 @@ module.exports = class StatusQueue {
-                            q.service.deleteMessage(q.name, m.messageId, m.popReceipt, (err) => {
-                                if (err) return callback(err);
-                            });
-                        }
-
-                        if (result.length == n) {
-                            return callback(null, { done: true, nonEmptyQs, result });
-                        }
+            const response = await q.service.receiveMessages({ numOfMessages: n });
+            let messages = response.receivedMessageItems;
+            for (let m of messages) {
+                if (m && Object.keys(m).length > 0) {
+                    result.push(options && options.raw ? m : this.deserializeMessage(m));
+
+                    if (!(options && options.remove === false)) {
+                        q.service.deleteMessage(m.messageId, m.popReceipt);
+                    }
+                    if (result.length == n) {
+                        return { done: true, nonEmptyQs, result };
                     }
                 }
-
-                if (i == qs.length - 1) {
-                    return callback(null, { done: nonEmptyQs.length === 0, nonEmptyQs, result });
-                }
-            });
+            }
         }
+        return { done: nonEmptyQs.length === 0, nonEmptyQs, result };
     }
 
-    pop(callback, n = 1, options = null) {
-        return this.getQueuesFunc((err, queues) => {
-            if (err) return callback(err);
-
-            const qServices = shuffle(this._getQServices(queues));
-            const perQ = qServices.length > 1 ? Math.floor(n / qServices.length) : qServices.length;
-
-            // first, iterate evenly and randomly on status queues
-            return this._pop(qServices, perQ, options, (err, partial) => {
-                if (err) return callback(err);
-
-                if (partial.done) return callback(null, partial.result);
 
-                let messagesLeftToPop = n - partial.result.length;
+    async pop(n = 1, options = null) {
+        const queues = await this.getQueuesFunc();
+        const qServices = shuffle(this._getQServices(queues));
+        const perQ = qServices.length > 1 ? Math.floor(n / qServices.length) : qServices.length;
 
-                // incase queues are uneven, iterate again, this time, request entire n messages, and trim
-                return this._pop(partial.result.nonEmptyQs, messagesLeftToPop, options, (err, final) => {
-                    if (err) return callback(err);
+        // first, iterate evenly and randomly on status queues
+        const partial = await this._pop(qServices, perQ, options);
+        if (partial.done) return partial.result;
 
-                    return callback(null, partial.result.concat(final.result));
-                });
-            });
+        let messagesLeftToPop = n - partial.result.length;
 
-        });
+        // incase queues are uneven, iterate again, this time, request entire n messages, and trim
```

Comment:
// In case queues are uneven, iterate again. This time, request for all n messages and trim